File: articles/applied-ai-services/form-recognizer/prebuilt-models/sample-code/csharp/prebuilt-tax.us.w2.2022-08-31.cs
using System;
using Azure;
using Azure.AI.FormRecognizer.DocumentAnalysis;

namespace FormRecognizerSample
{
    internal class Program
    {
        static void Main(string[] args)
        {
            var endpoint = args.Length > 0 ? args[0] : "<your-endpoint>";
            var key = args.Length > 1 ? args[1] : "<your-key>";
            var documentUrl = args.Length > 2 ? args[2] : "<your-document-url>";
            var modelId = "prebuilt-tax.us.w2";

            var client = new DocumentAnalysisClient(new Uri(endpoint), new AzureKeyCredential(key));
            var operation = client.StartAnalyzeDocumentFromUri(modelId, new Uri(documentUrl));
            var result = operation.WaitForCompletion().Value;
            foreach (var document in result.Documents)
            {
                switch (document.DocType)
                {
                    case "tax.us.w2": ProcessTax_Us_W2(document); break;
                }
            }
        }

        static void ProcessTax_Us_W2(AnalyzedDocument document)
        {
            Console.WriteLine($"Document: {document.DocType}");
            if (document.Fields.TryGetValue("W2FormVariant", out DocumentField? w2FormVariantField))
            {
                Console.WriteLine($"  W2FormVariant: Value={w2FormVariantField.AsString()}   Confidence={w2FormVariantField.Confidence}");
            }
            if (document.Fields.TryGetValue("TaxYear", out DocumentField? taxYearField))
            {
                Console.WriteLine($"  TaxYear: Value={taxYearField.AsString()}   Confidence={taxYearField.Confidence}");
            }
            if (document.Fields.TryGetValue("W2Copy", out DocumentField? w2CopyField))
            {
                Console.WriteLine($"  W2Copy: Value={w2CopyField.AsString()}   Confidence={w2CopyField.Confidence}");
            }
            if (document.Fields.TryGetValue("Employee", out DocumentField? employeeField))
            {
                var employeeFieldFields = employeeField.AsDictionary();
                if (employeeFieldFields.TryGetValue("SocialSecurityNumber", out DocumentField? socialSecurityNumberField))
                {
                    Console.WriteLine($"  Employee.SocialSecurityNumber: Value={socialSecurityNumberField.AsString()}   Confidence={socialSecurityNumberField.Confidence}");
                }
                if (employeeFieldFields.TryGetValue("Name", out DocumentField? nameField))
                {
                    Console.WriteLine($"  Employee.Name: Value={nameField.AsString()}   Confidence={nameField.Confidence}");
                }
                if (employeeFieldFields.TryGetValue("Address", out DocumentField? addressField))
                {
                    Console.WriteLine($"  Employee.Address: Value={addressField.AsAddress()}   Confidence={addressField.Confidence}");
                }
            }
            if (document.Fields.TryGetValue("ControlNumber", out DocumentField? controlNumberField))
            {
                Console.WriteLine($"  ControlNumber: Value={controlNumberField.AsString()}   Confidence={controlNumberField.Confidence}");
            }
            if (document.Fields.TryGetValue("Employer", out DocumentField? employerField))
            {
                var employerFieldFields = employerField.AsDictionary();
                if (employerFieldFields.TryGetValue("IdNumber", out DocumentField? idNumberField))
                {
                    Console.WriteLine($"  Employer.IdNumber: Value={idNumberField.AsString()}   Confidence={idNumberField.Confidence}");
                }
                if (employerFieldFields.TryGetValue("Name", out DocumentField? nameField))
                {
                    Console.WriteLine($"  Employer.Name: Value={nameField.AsString()}   Confidence={nameField.Confidence}");
                }
                if (employerFieldFields.TryGetValue("Address", out DocumentField? addressField))
                {
                    Console.WriteLine($"  Employer.Address: Value={addressField.AsAddress()}   Confidence={addressField.Confidence}");
                }
            }
            if (document.Fields.TryGetValue("WagesTipsAndOtherCompensation", out DocumentField? wagesTipsAndOtherCompensationField))
            {
                Console.WriteLine($"  WagesTipsAndOtherCompensation: Value={wagesTipsAndOtherCompensationField.AsDouble()}   Confidence={wagesTipsAndOtherCompensationField.Confidence}");
            }
            if (document.Fields.TryGetValue("FederalIncomeTaxWithheld", out DocumentField? federalIncomeTaxWithheldField))
            {
                Console.WriteLine($"  FederalIncomeTaxWithheld: Value={federalIncomeTaxWithheldField.AsDouble()}   Confidence={federalIncomeTaxWithheldField.Confidence}");
            }
            if (document.Fields.TryGetValue("SocialSecurityWages", out DocumentField? socialSecurityWagesField))
            {
                Console.WriteLine($"  SocialSecurityWages: Value={socialSecurityWagesField.AsDouble()}   Confidence={socialSecurityWagesField.Confidence}");
            }
            if (document.Fields.TryGetValue("SocialSecurityTaxWithheld", out DocumentField? socialSecurityTaxWithheldField))
            {
                Console.WriteLine($"  SocialSecurityTaxWithheld: Value={socialSecurityTaxWithheldField.AsDouble()}   Confidence={socialSecurityTaxWithheldField.Confidence}");
            }
            if (document.Fields.TryGetValue("MedicareWagesAndTips", out DocumentField? medicareWagesAndTipsField))
            {
                Console.WriteLine($"  MedicareWagesAndTips: Value={medicareWagesAndTipsField.AsDouble()}   Confidence={medicareWagesAndTipsField.Confidence}");
            }
            if (document.Fields.TryGetValue("MedicareTaxWithheld", out DocumentField? medicareTaxWithheldField))
            {
                Console.WriteLine($"  MedicareTaxWithheld: Value={medicareTaxWithheldField.AsDouble()}   Confidence={medicareTaxWithheldField.Confidence}");
            }
            if (document.Fields.TryGetValue("SocialSecurityTips", out DocumentField? socialSecurityTipsField))
            {
                Console.WriteLine($"  SocialSecurityTips: Value={socialSecurityTipsField.AsDouble()}   Confidence={socialSecurityTipsField.Confidence}");
            }
            if (document.Fields.TryGetValue("AllocatedTips", out DocumentField? allocatedTipsField))
            {
                Console.WriteLine($"  AllocatedTips: Value={allocatedTipsField.AsDouble()}   Confidence={allocatedTipsField.Confidence}");
            }
            if (document.Fields.TryGetValue("VerificationCode", out DocumentField? verificationCodeField))
            {
                Console.WriteLine($"  VerificationCode: Value={verificationCodeField.AsString()}   Confidence={verificationCodeField.Confidence}");
            }
            if (document.Fields.TryGetValue("DependentCareBenefits", out DocumentField? dependentCareBenefitsField))
            {
                Console.WriteLine($"  DependentCareBenefits: Value={dependentCareBenefitsField.AsDouble()}   Confidence={dependentCareBenefitsField.Confidence}");
            }
            if (document.Fields.TryGetValue("NonQualifiedPlans", out DocumentField? nonQualifiedPlansField))
            {
                Console.WriteLine($"  NonQualifiedPlans: Value={nonQualifiedPlansField.AsDouble()}   Confidence={nonQualifiedPlansField.Confidence}");
            }
            if (document.Fields.TryGetValue("AdditionalInfo", out DocumentField? additionalInfoField))
            {
                var index = 0;
                foreach (var item in additionalInfoField.AsList())
                {
                    var itemFields = item.AsDictionary();
                    if (itemFields.TryGetValue("LetterCode", out DocumentField? letterCodeField))
                    {
                        Console.WriteLine($"  AdditionalInfo[{index}].LetterCode: Value={letterCodeField.AsString()}   Confidence={letterCodeField.Confidence}");
                    }
                    if (itemFields.TryGetValue("Amount", out DocumentField? amountField))
                    {
                        Console.WriteLine($"  AdditionalInfo[{index}].Amount: Value={amountField.AsDouble()}   Confidence={amountField.Confidence}");
                    }
                    index++;
                }
            }
            if (document.Fields.TryGetValue("IsStatutoryEmployee", out DocumentField? isStatutoryEmployeeField))
            {
                Console.WriteLine($"  IsStatutoryEmployee: Value={isStatutoryEmployeeField.AsString()}   Confidence={isStatutoryEmployeeField.Confidence}");
            }
            if (document.Fields.TryGetValue("IsRetirementPlan", out DocumentField? isRetirementPlanField))
            {
                Console.WriteLine($"  IsRetirementPlan: Value={isRetirementPlanField.AsString()}   Confidence={isRetirementPlanField.Confidence}");
            }
            if (document.Fields.TryGetValue("IsThirdPartySickPay", out DocumentField? isThirdPartySickPayField))
            {
                Console.WriteLine($"  IsThirdPartySickPay: Value={isThirdPartySickPayField.AsString()}   Confidence={isThirdPartySickPayField.Confidence}");
            }
            if (document.Fields.TryGetValue("Other", out DocumentField? otherField))
            {
                Console.WriteLine($"  Other: Value={otherField.AsString()}   Confidence={otherField.Confidence}");
            }
            if (document.Fields.TryGetValue("StateTaxInfos", out DocumentField? stateTaxInfosField))
            {
                var index = 0;
                foreach (var item in stateTaxInfosField.AsList())
                {
                    var itemFields = item.AsDictionary();
                    if (itemFields.TryGetValue("State", out DocumentField? stateField))
                    {
                        Console.WriteLine($"  StateTaxInfos[{index}].State: Value={stateField.AsString()}   Confidence={stateField.Confidence}");
                    }
                    if (itemFields.TryGetValue("EmployerStateIdNumber", out DocumentField? employerStateIdNumberField))
                    {
                        Console.WriteLine($"  StateTaxInfos[{index}].EmployerStateIdNumber: Value={employerStateIdNumberField.AsString()}   Confidence={employerStateIdNumberField.Confidence}");
                    }
                    if (itemFields.TryGetValue("StateWagesTipsEtc", out DocumentField? stateWagesTipsEtcField))
                    {
                        Console.WriteLine($"  StateTaxInfos[{index}].StateWagesTipsEtc: Value={stateWagesTipsEtcField.AsDouble()}   Confidence={stateWagesTipsEtcField.Confidence}");
                    }
                    if (itemFields.TryGetValue("StateIncomeTax", out DocumentField? stateIncomeTaxField))
                    {
                        Console.WriteLine($"  StateTaxInfos[{index}].StateIncomeTax: Value={stateIncomeTaxField.AsDouble()}   Confidence={stateIncomeTaxField.Confidence}");
                    }
                    index++;
                }
            }
            if (document.Fields.TryGetValue("LocalTaxInfos", out DocumentField? localTaxInfosField))
            {
                var index = 0;
                foreach (var item in localTaxInfosField.AsList())
                {
                    var itemFields = item.AsDictionary();
                    if (itemFields.TryGetValue("LocalWagesTipsEtc", out DocumentField? localWagesTipsEtcField))
                    {
                        Console.WriteLine($"  LocalTaxInfos[{index}].LocalWagesTipsEtc: Value={localWagesTipsEtcField.AsDouble()}   Confidence={localWagesTipsEtcField.Confidence}");
                    }
                    if (itemFields.TryGetValue("LocalIncomeTax", out DocumentField? localIncomeTaxField))
                    {
                        Console.WriteLine($"  LocalTaxInfos[{index}].LocalIncomeTax: Value={localIncomeTaxField.AsDouble()}   Confidence={localIncomeTaxField.Confidence}");
                    }
                    if (itemFields.TryGetValue("LocalityName", out DocumentField? localityNameField))
                    {
                        Console.WriteLine($"  LocalTaxInfos[{index}].LocalityName: Value={localityNameField.AsString()}   Confidence={localityNameField.Confidence}");
                    }
                    index++;
                }
            }
        }
    }
}

File: articles/applied-ai-services/form-recognizer/prebuilt-models/sample-code/python/prebuilt-tax.us.w2.2022-08-31.py
import os
import sys
from azure.core.credentials import AzureKeyCredential
from azure.ai.formrecognizer import DocumentAnalysisClient

def analyze_document(args):
    endpoint = args[1] if len(args) > 1 else "<your-endpoint>"
    key = args[2] if len(args) > 2 else "<your-key>"
    doc_url = args[3] if len(args) > 3 else "<your-document-url>"
    model_id = "prebuilt-tax.us.w2";

    document_analysis_client = DocumentAnalysisClient(endpoint=endpoint, credential=AzureKeyCredential(key))
    poller = document_analysis_client.begin_analyze_document_from_url(model_id, doc_url)
    result = poller.result()
    for document in result.documents:
        if document.doc_type == "tax.us.w2":
            process_tax_us_w_(document)


def process_tax_us_w_(document):
    print(f"Document: {document.doc_type}")

    w__form_variant_field = document.fields.get("W2FormVariant")
    if w__form_variant_field:
        print(f"    W2FormVariant: Value={w__form_variant_field.value}   Confidence={w__form_variant_field.confidence}")
    
    tax_year_field = document.fields.get("TaxYear")
    if tax_year_field:
        print(f"    TaxYear: Value={tax_year_field.value}   Confidence={tax_year_field.confidence}")
    
    w__copy_field = document.fields.get("W2Copy")
    if w__copy_field:
        print(f"    W2Copy: Value={w__copy_field.value}   Confidence={w__copy_field.confidence}")
    
    employee_field = document.fields.get("Employee")
    if employee_field:
        employee_field_fields = employee_field.value
        
        social_security_number_field = employee_field_fields.get("SocialSecurityNumber")
        if social_security_number_field:
            print(f"    Employee.SocialSecurityNumber: Value={social_security_number_field.value}   Confidence={social_security_number_field.confidence}")
        
        name_field = employee_field_fields.get("Name")
        if name_field:
            print(f"    Employee.Name: Value={name_field.value}   Confidence={name_field.confidence}")
        
        address_field = employee_field_fields.get("Address")
        if address_field:
            print(f"    Employee.Address: Value={address_field.value}   Confidence={address_field.confidence}")
        
    
    control_number_field = document.fields.get("ControlNumber")
    if control_number_field:
        print(f"    ControlNumber: Value={control_number_field.value}   Confidence={control_number_field.confidence}")
    
    employer_field = document.fields.get("Employer")
    if employer_field:
        employer_field_fields = employer_field.value
        
        id_number_field = employer_field_fields.get("IdNumber")
        if id_number_field:
            print(f"    Employer.IdNumber: Value={id_number_field.value}   Confidence={id_number_field.confidence}")
        
        name_field = employer_field_fields.get("Name")
        if name_field:
            print(f"    Employer.Name: Value={name_field.value}   Confidence={name_field.confidence}")
        
        address_field = employer_field_fields.get("Address")
        if address_field:
            print(f"    Employer.Address: Value={address_field.value}   Confidence={address_field.confidence}")
        
    
    wages_tips_and_other_compensation_field = document.fields.get("WagesTipsAndOtherCompensation")
    if wages_tips_and_other_compensation_field:
        print(f"    WagesTipsAndOtherCompensation: Value={wages_tips_and_other_compensation_field.value}   Confidence={wages_tips_and_other_compensation_field.confidence}")
    
    federal_income_tax_withheld_field = document.fields.get("FederalIncomeTaxWithheld")
    if federal_income_tax_withheld_field:
        print(f"    FederalIncomeTaxWithheld: Value={federal_income_tax_withheld_field.value}   Confidence={federal_income_tax_withheld_field.confidence}")
    
    social_security_wages_field = document.fields.get("SocialSecurityWages")
    if social_security_wages_field:
        print(f"    SocialSecurityWages: Value={social_security_wages_field.value}   Confidence={social_security_wages_field.confidence}")
    
    social_security_tax_withheld_field = document.fields.get("SocialSecurityTaxWithheld")
    if social_security_tax_withheld_field:
        print(f"    SocialSecurityTaxWithheld: Value={social_security_tax_withheld_field.value}   Confidence={social_security_tax_withheld_field.confidence}")
    
    medicare_wages_and_tips_field = document.fields.get("MedicareWagesAndTips")
    if medicare_wages_and_tips_field:
        print(f"    MedicareWagesAndTips: Value={medicare_wages_and_tips_field.value}   Confidence={medicare_wages_and_tips_field.confidence}")
    
    medicare_tax_withheld_field = document.fields.get("MedicareTaxWithheld")
    if medicare_tax_withheld_field:
        print(f"    MedicareTaxWithheld: Value={medicare_tax_withheld_field.value}   Confidence={medicare_tax_withheld_field.confidence}")
    
    social_security_tips_field = document.fields.get("SocialSecurityTips")
    if social_security_tips_field:
        print(f"    SocialSecurityTips: Value={social_security_tips_field.value}   Confidence={social_security_tips_field.confidence}")
    
    allocated_tips_field = document.fields.get("AllocatedTips")
    if allocated_tips_field:
        print(f"    AllocatedTips: Value={allocated_tips_field.value}   Confidence={allocated_tips_field.confidence}")
    
    verification_code_field = document.fields.get("VerificationCode")
    if verification_code_field:
        print(f"    VerificationCode: Value={verification_code_field.value}   Confidence={verification_code_field.confidence}")
    
    dependent_care_benefits_field = document.fields.get("DependentCareBenefits")
    if dependent_care_benefits_field:
        print(f"    DependentCareBenefits: Value={dependent_care_benefits_field.value}   Confidence={dependent_care_benefits_field.confidence}")
    
    non_qualified_plans_field = document.fields.get("NonQualifiedPlans")
    if non_qualified_plans_field:
        print(f"    NonQualifiedPlans: Value={non_qualified_plans_field.value}   Confidence={non_qualified_plans_field.confidence}")
    
    additional_info_field = document.fields.get("AdditionalInfo")
    if additional_info_field:
        for index, item in enumerate(additional_info_field.value):
            item_fields = item.value
            
            letter_code_field = item_fields.get("LetterCode")
            if letter_code_field:
                print(f"    AdditionalInfo[{index}].LetterCode: Value={letter_code_field.value}   Confidence={letter_code_field.confidence}")
            
            amount_field = item_fields.get("Amount")
            if amount_field:
                print(f"    AdditionalInfo[{index}].Amount: Value={amount_field.value}   Confidence={amount_field.confidence}")
            
    
    is_statutory_employee_field = document.fields.get("IsStatutoryEmployee")
    if is_statutory_employee_field:
        print(f"    IsStatutoryEmployee: Value={is_statutory_employee_field.value}   Confidence={is_statutory_employee_field.confidence}")
    
    is_retirement_plan_field = document.fields.get("IsRetirementPlan")
    if is_retirement_plan_field:
        print(f"    IsRetirementPlan: Value={is_retirement_plan_field.value}   Confidence={is_retirement_plan_field.confidence}")
    
    is_third_party_sick_pay_field = document.fields.get("IsThirdPartySickPay")
    if is_third_party_sick_pay_field:
        print(f"    IsThirdPartySickPay: Value={is_third_party_sick_pay_field.value}   Confidence={is_third_party_sick_pay_field.confidence}")
    
    other_field = document.fields.get("Other")
    if other_field:
        print(f"    Other: Value={other_field.value}   Confidence={other_field.confidence}")
    
    state_tax_infos_field = document.fields.get("StateTaxInfos")
    if state_tax_infos_field:
        for index, item in enumerate(state_tax_infos_field.value):
            item_fields = item.value
            
            state_field = item_fields.get("State")
            if state_field:
                print(f"    StateTaxInfos[{index}].State: Value={state_field.value}   Confidence={state_field.confidence}")
            
            employer_state_id_number_field = item_fields.get("EmployerStateIdNumber")
            if employer_state_id_number_field:
                print(f"    StateTaxInfos[{index}].EmployerStateIdNumber: Value={employer_state_id_number_field.value}   Confidence={employer_state_id_number_field.confidence}")
            
            state_wages_tips_etc_field = item_fields.get("StateWagesTipsEtc")
            if state_wages_tips_etc_field:
                print(f"    StateTaxInfos[{index}].StateWagesTipsEtc: Value={state_wages_tips_etc_field.value}   Confidence={state_wages_tips_etc_field.confidence}")
            
            state_income_tax_field = item_fields.get("StateIncomeTax")
            if state_income_tax_field:
                print(f"    StateTaxInfos[{index}].StateIncomeTax: Value={state_income_tax_field.value}   Confidence={state_income_tax_field.confidence}")
            
    
    local_tax_infos_field = document.fields.get("LocalTaxInfos")
    if local_tax_infos_field:
        for index, item in enumerate(local_tax_infos_field.value):
            item_fields = item.value
            
            local_wages_tips_etc_field = item_fields.get("LocalWagesTipsEtc")
            if local_wages_tips_etc_field:
                print(f"    LocalTaxInfos[{index}].LocalWagesTipsEtc: Value={local_wages_tips_etc_field.value}   Confidence={local_wages_tips_etc_field.confidence}")
            
            local_income_tax_field = item_fields.get("LocalIncomeTax")
            if local_income_tax_field:
                print(f"    LocalTaxInfos[{index}].LocalIncomeTax: Value={local_income_tax_field.value}   Confidence={local_income_tax_field.confidence}")
            
            locality_name_field = item_fields.get("LocalityName")
            if locality_name_field:
                print(f"    LocalTaxInfos[{index}].LocalityName: Value={locality_name_field.value}   Confidence={locality_name_field.confidence}")
            
    
if __name__ == "__main__":
    analyze_document(sys.argv)